synapse/rest/client/login.py

```diff
@@ -413,7 +414,8 @@ async def _do_jwt_login(
                 errcode=Codes.FORBIDDEN,
             )
 
-        user = payload.get("sub", None)
+        subject_claim = self.jwt_subject_claim or "sub"
+        user = payload.get(subject_claim, None)
```

Comment:
Overall this looks reasonable (I think it is based on a slimmed down version of #9493). I appreciate splitting it up a bit!

My only comment is that there shouldn't be any reason to include the default `"sub"` again here. I think the above code should be:

```suggestion
        user = payload.get(self.jwt_subject_claim, None)
```

Reply:
Ah, yeah, I recall this from https://github.com/matrix-org/synapse/pull/9493#discussion_r660899401. I'll try your suggestion out once I get the tests running locally.

Reply:
@clokep with your suggested change, I am seeing a few test failures:
```
[nix-shell:~/src/github.com/matrix-org/synapse]$ trial tests/rest/client/test_login.py | grep -v '\[OK\]'
tests.rest.client.test_login
  JWTPubKeyTestCase
    test_login_jwt_valid ...                                             [FAIL]
  JWTTestCase
    test_login_default_sub ...                                           [FAIL]
    test_login_iss_no_config ...                                         [FAIL]
    test_login_jwt_valid_registered ...                                  [FAIL]
    test_login_jwt_valid_unregistered ...                                [FAIL]
```

I assume something is not set up the way it should to pick the default value up it these tests. Any tips how to fix this, or where to find a good example to follow?

Reply:
@vrinek The `JWTPubKeyTestCase` and `JWTTestCase` should be updated to define `default_config` instead of touching the homeserver object in `make_homeserver`. It would be something like:

```diff
diff --git a/tests/rest/client/test_login.py b/tests/rest/client/test_login.py
--- a/tests/rest/client/test_login.py
+++ b/tests/rest/client/test_login.py
@@ -1024,12 +1024,14 @@ class JWTPubKeyTestCase(unittest.HomeserverTestCase):
         ]
     )
 
-    def make_homeserver(self, reactor, clock):
-        self.hs = self.setup_test_homeserver()
-        self.hs.config.jwt.jwt_enabled = True
-        self.hs.config.jwt.jwt_secret = self.jwt_pubkey
-        self.hs.config.jwt.jwt_algorithm = "RS256"
-        return self.hs
+    def default_config(self):
+        config = super().default_config()
+        config["jwt_config"] = {
+            "enabled": True,
+            "secret": self.jwt_pubkey,
+            "algorithm": "RS256",
+        }
+        return config
 
     def jwt_encode(self, payload: Dict[str, Any], secret: str = jwt_privatekey) -> str:
         # PyJWT 2.0.0 changed the return type of jwt.encode from bytes to str.
```

I'm not sure if that will work exactly, but hopefully it will! This is a more robust way of setting up a homeserver during testing!

Reply:
@clokep thanks for the code snippet. It works for the majority of test but fails for the ones with an `@override_config`.

I am trying to debug this now, but any further advice will be greatly appreciated.

Reply:
Ok, I figured something out: `default_config` & `@override_config` don't play well together:

https://github.com/MQSdk/synapse/blob/1ff87641192d61a05756728c7be2503808442e06/tests/unittest.py#L385-L396
```python
    def default_config(self):
        """[...]"""
        config = default_config("test")

        # apply any additional config which was specified via the override_config
        # decorator.
        if self._extra_config is not None:
            config.update(self._extra_config)

        return config
```

https://github.com/MQSdk/synapse/blob/1ff87641192d61a05756728c7be2503808442e06/tests/unittest.py#L795-L816
```python
def override_config(extra_config):
    """[...]"""

    def decorator(func):
        func._extra_config = extra_config
        return func

    return decorator
```

Our `default_config` override:
```python

    def default_config(self):
        config = super().default_config()
        config["jwt_config"] = {
            "enabled": True,
            "secret": self.jwt_secret,
            "algorithm": self.jwt_algorithm,
        }
```

As far as I understand, our `default_config` will apply the `@override_config` and then overwrite it. I couldn't find a great way to circumvent this. My only good enough example going forward is from https://github.com/MQSdk/synapse/tree/1ff87641192d61a05756728c7be2503808442e06/tests/handlers/test_oidc.py which uses a `DEFAULT_CONFIG` and then updates it for every test that it needs to overwrite parts of it.

Reply:
@clokep I managed to fix it up. Please give it another review and let me know of any thing else that needs work.